Implement Plutarch Rationals

Geometer1729 opened the pull request:
I'm not certain this is the best way to do Rationals in Plutarch, but we're going to need them fairly soon for Danaswap and as long as any future improvements can be backwards compatible with the Num and Fractional instances this should at least allow us to write and test code using rationals for now.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -1,28 +1,118 @@
+module Plutarch.Rational 
+  (PRational
+  ,preduce
+  ,pnumerator
```
./bin/format

nix build .#check.x86_64-linux

L-as commented on Plutarch/Rational.hs:
```diff
@@ -1,28 +1,118 @@
+    pcon $ PRational (fromInteger $ numerator r) (fromInteger $ denominator r)
+
+preduce :: Term s (PRational :--> PRational)
+preduce = plam $ \x ->
```
You need to hoist all your terms.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -1,28 +1,118 @@
+      pcon (PRational xd xn))
+  x / y = preduce #
+    pmatch x (\(PRational xn xd) ->
+    pmatch y (\(PRational yn yd) ->
```
Please do `pmatch x $ ...`

L-as commented on Plutarch/Rational.hs:
```diff
@@ -1,28 +1,118 @@
+            a
+            (self # (pmod # a # b) # b)
+
+pmin,pmax :: POrd a => Term s (a :--> a :--> a)
```
Please don't declare multiple functions together. See github.com/mlabs-haskell/styleguide

L-as reviewed (changes requested):
Waiting for the review of @kozross

Geometer1729 commented:
The arithmetic terms don't seem to be hoist-able because they are not term level functions. `pgcd'`,`pmin` and `pmax` are all used just once. For `pnumerator` , `pdenominator` and `pfromInteger` I'm not sure if I should hoist them. They are all pretty simple.

srid commented:
I think we are also going to need a `PIsData` instance, to be able to store rationals in datum/redeemer.

L-as commented:
> pgcd',pmin and pmax are all used just once. For pnumerator , pdenominator and pfromInteger I'm not sure if I should hoist them. They are all pretty simple.

The rule of thumb is to hoist everything you can. If it's used only once, then it will be inlined by the compiler.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -1,28 +1,125 @@
+
+instance POrd PRational where
+  l #<= r =
+    pmatch l $ \(PRational ln ld) ->
```
@TotallyNotChase this should probably also be in the guide

L-as replied:
#92

kozross reviewed (dismissed):
Looks good to me. Are there plans to add `round`, `truncate` and/or `properFraction` support?

L-as commented:
Needs to merge in upstream.

L-as commented:
> Looks good to me. Are there plans to add `round`, `truncate` and/or `properFraction` support?

If you want to add them, sure.

L-as reviewed (dismissed):
LGTM

Geometer1729 commented:
@srid I tried to write a PIsData instance but I'm not really sure if this is right.

Geometer1729 commented:
>

Geometer1729 commented:
I'm sorry I clicked the wrong button  and closed this.

Geometer1729 commented:
> Looks good to me. Are there plans to add `round`, `truncate` and/or `properFraction` support?

I kinda forgot about those. Looks like we do need them. I'll start working on that.

srid commented:
> @srid I tried to write a PIsData instance but I'm not really sure if this is right.

You can add a test to `examples/Examples/Rationals.hs` to verify that `pfromData . pdata` is essentially `id`, on a bunch of inputs (property tests would be better, but we don't use them yet).

Geometer1729 commented:
@srid that test revealed a bug, I fixed it and now the tests pass, hopefully that was the only bug.
@kozross I implemented the rounding functions you mentioned.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -1,3 +1,4 @@
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
```
?

L-as replied:
I don't think we need the other pragma either.

kozross replied:
This is a _really_ bad thing to enable in general.

Geometer1729 replied:
Sorry about that, I think the haskell-language-server plugin I'm trying out added those. I'm not sure why. I've removed both pragmas.

L-as commented on Plutarch/Rational.hs:
```diff
@@ -11,12 +12,33 @@ module Plutarch.Rational (
+  pfromData x = pListToRat #$ pmap # pasInt #$ pasList # pforgetData x
+  pdata x =
+    (punsafeCoerce :: Term _ (PAsData (PBuiltinList (PAsData PInteger))) -> Term _ (PAsData PRational)) $
+      pdata $ pmap # plam pdata #$ pRatToList # x
```
Why not just do `pdata $` instead of `plam pdata #$`?

L-as replied:
Otherwise this looks fine actually :+1: , however, there is a design choice to be made here, between this approach and another approach. @TotallyNotChase

L-as replied:
#56

L-as replied:
I just noticed these aren't hoisted.

L-as commented:
In general, I recommend not adding more features to PRs. You should just make a new PR.

kozross commented on Plutarch/Rational.hs:
```diff
@@ -1,28 +1,235 @@
+pfromInteger :: Term s (PInteger :--> PRational)
+pfromInteger = phoistAcyclic $ plam $ \n -> pcon $ PRational n 1
+
+pround :: Term s (PRational :--> PInteger)
```
Does this match the behaviour of `round` from both Plutus and `base`? Specifically:

``round x`` returns the nearest integer to ``x``; the even integer if ``x`` is equidistant between two integers

Geometer1729 replied:
It does not I didn't realize that was the behavior of round. I'll fix that.

kozross commented on examples/Examples/Rationals.hs:
```diff
@@ -0,0 +1,65 @@
+        expect $ 1 / 2 + 1 / 3 + 1 / 4 + 1 / 5 #== (77 / 60 :: Term s PRational)
+    , testCase "product" $
+        expect $ 1 / 2 * 2 / 3 * 3 / 4 * 4 / 5 * 5 / 6 #== (1 / 6 :: Term s PRational)
+    , testCase "round 5/3" $
+        expect $ pround # (5 / 3) #== 2
+    , testCase "round 4/3" $
+        expect $ pround # (4 / 3) #== 1
+    , testCase "round 5/2" $
+        expect $ pround # (5 / 2) #== 3
+    , testCase "round 9/4" $
+        expect $ pround # (9 / 4) #== 2
+    , testCase "round 11/4" $
+        expect $ pround # (11 / 4) #== 3
```
We need to cover the 'equidistant' case described above. There are also no tests dealing with negative values.

kozross commented on examples/Examples/Rationals.hs:
```diff
@@ -0,0 +1,65 @@
+        expect $ pround # (9 / 4) #== 2
+    , testCase "round 11/4" $
+        expect $ pround # (11 / 4) #== 3
+    , testCase "truncate 5/4" $
+        expect $ ptruncate # (5 / 4) #== 1
+    , testCase "truncate 3/2" $
+        expect $ ptruncate # (3 / 2) #== 1
+    , testCase "truncate 7/4" $
+        expect $ ptruncate # (7 / 4) #== 1
+    , testCase "properFraction 11/7" $
+        expect $
+          pmatch (pproperFraction # (11 / 7)) $ \(PPair x y) ->
+            x #== 1 #&& y #== (4 / 7)
+    , testCase "properFraction 13/7" $
+        expect $
+          pmatch (pproperFraction # (13 / 7)) $ \(PPair x y) ->
+            x #== 1 #&& y #== (6 / 7)
```
Similarly here: we need to at least deal with negative values properly.

L-as commented:
Please retarget to staging BTW>

Geometer1729 commented:
> In general, I recommend not adding more features to PRs. You should just make a new PR.

Sorry about that, I'll keep that in mind going forward.

Geometer1729 commented:
I think I've made all of the requested changes now.

L-as commented:
This has merge conflicts now unfortunately.

L-as commented:
I think I solved it using GH's built-in functionality.

L-as commented:
Formatting is wrong FWIW. `./bin/format` needs to be run. CI is broken because staging branch is broken.
